data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/S3SinkService.java

```diff
@@ -0,0 +1,172 @@
+    public static final String NUMBER_OF_RECORDS_FLUSHED_TO_S3_SUCCESS = "numberOfRecordsFlushedToS3Success";
+    public static final String NUMBER_OF_RECORDS_FLUSHED_TO_S3_FAILED = "numberOfRecordsFlushedToS3Failed";
+    static final String S3_OBJECTS_SIZE = "s3ObjectSizeRecords";
+    private final S3SinkConfig s3SinkConfig;
+    private final Lock reentrantLock;
+    private final BufferFactory bufferFactory;
+    private final Codec codec;
+    private Buffer currentBuffer;
+    private final int maxEvents;
+    private final ByteCount maxBytes;
+    private final long maxCollectionDuration;
+    private final String bucket;
+    private final int maxRetries;
+    private final Counter snapshotSuccessCounter;
+    private final Counter snapshotFailedCounter;
+    private final Counter numberOfRecordsSuccessCounter;
+    private final Counter numberOfRecordsFailedCounter;
+    private final DistributionSummary s3ObjectSizeSummary;
+
+    /**
+     * @param s3SinkConfig  s3 sink related configuration.
+     * @param codec         parser
+     * @param pluginMetrics metrics
+     */
+    public S3SinkService(final S3SinkConfig s3SinkConfig, final BufferFactory bufferFactory,
+                         final Codec codec, PluginMetrics pluginMetrics) {
+        this.s3SinkConfig = s3SinkConfig;
+        this.bufferFactory = bufferFactory;
+        this.codec = codec;
+        reentrantLock = new ReentrantLock();
+
+        maxEvents = s3SinkConfig.getThresholdOptions().getEventCount();
+        maxBytes = s3SinkConfig.getThresholdOptions().getMaximumSize();
+        maxCollectionDuration = s3SinkConfig.getThresholdOptions().getEventCollectTimeOut().getSeconds();
+
+        bucket = s3SinkConfig.getBucketOptions().getBucketName();
+        maxRetries = s3SinkConfig.getMaxUploadRetries();
+
+        snapshotSuccessCounter = pluginMetrics.counter(SNAPSHOT_SUCCESS);
+        snapshotFailedCounter = pluginMetrics.counter(SNAPSHOT_FAILED);
+        numberOfRecordsSuccessCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_FLUSHED_TO_S3_SUCCESS);
+        numberOfRecordsFailedCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_FLUSHED_TO_S3_FAILED);
+        s3ObjectSizeSummary = pluginMetrics.summary(S3_OBJECTS_SIZE);
+    }
+
+    /**
+     * @param records received records and add into buffer.
+     */
+    void output(Collection<Record<Event>> records) {
+        reentrantLock.lock();
+        if (currentBuffer == null) {
+            currentBuffer = bufferFactory.getBuffer();
+        }
+        try {
+            for (Record<Event> record : records) {
+
+                final Event event = record.getData();
+                final String encodedEvent;
+                encodedEvent = codec.parse(event);
+                final byte[] encodedBytes = encodedEvent.getBytes();
+
+                if (ThresholdCheck.checkThresholdExceed(currentBuffer, maxEvents, maxBytes, maxCollectionDuration)) {
```

Comment:
If the maxEvents is 10 and only one event is received, no other event is sent for one hour, I don't see how that one event is flushed to S3.

Reply:
This is handled in event collection duration i.e. event_collect_timeout .  When the duration is met it will be flushed to S3

Reply:
As mentioned in the email, this is not handled today. We need a way to handle this in all sinks. We will tackle this in a separate PR (either sink specific or at data prepper core level)